users_ldap_groups/tests/test_users_ldap_groups.py

```diff
@@ -0,0 +1,109 @@
+# Copyright 2018 Therp BV <https://therp.nl>
+# Copyright 2018 Brainbean Apps <https://brainbeanapps.com>
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+from unittest import mock
+from odoo.tests.common import TransactionCase
+from odoo import api, registry, SUPERUSER_ID
+
+_auth_ldap_ns = 'odoo.addons.auth_ldap'
+_company_ldap_class = _auth_ldap_ns + '.models.res_company_ldap.CompanyLDAP'
+
+
+class FakeLdapConnection(object):
+    def __init__(self, entries):
+        self.entries = entries
+
+    def simple_bind_s(self, dn, passwd):
+        pass
+
+    def search_st(self, dn, scope, ldap_filter, attributes, timeout):
+        if dn in self.entries:
+            return [(dn, dict(self.entries[dn]))]
+        return []
+
+    def unbind(self):
+        pass
+
+    def unbind_s(self):
+        pass
+
+    def __getattr__(self, name):
+        def wrapper():
+            raise Exception("'%s' is not mocked" % name)
+        return wrapper
+
+
+class TestUsersLdapGroups(TransactionCase):
+    def test_users_ldap_groups(self):
+        # _login does its work in a new cursor, so we need one too
+        with registry(self.env.cr.dbname).cursor() as cr:
+            env = api.Environment(cr, SUPERUSER_ID, {})
+
+            group_contains = env['res.groups'].create({'name': 'contains'})
+            group_equals = env['res.groups'].create({'name': 'equals'})
+            group_query = env['res.groups'].create({'name': 'query'})
+            env.ref('base.main_company').write({'ldaps': [
+                (5, False, False),
+                (0, 0, {
+                    'ldap_base': 'dc=users_ldap_groups,dc=example,dc=com',
+                    'ldap_filter': '(uid=%s)',
+                    'ldap_binddn': 'cn=bind,dc=example,dc=com',
+                    'create_user': True,
+                    'only_ldap_groups': True,
+                    'group_mapping_ids': [
+                        (0, 0, {
+                            'ldap_attribute': 'name',
+                            'operator': 'contains',
+                            'value': 'hello3',
+                            'group_id': env.ref('base.group_system').id,
+                        }),
```

Comment:
@alexey-pelykh IMO `env.ref('base.group_system')` use multipal time, so use like global variable

Reply:
If it would be a `self.env.ref` call - surely, yet in this case env is unique in each case, thus technically reference _may_ be different

Reply:
> If it would be a `self.env.ref` call - surely, yet in this case env is unique in each case, thus technically reference _may_ be different

@alexey-pelykh Then it's Ok

Reply:
Done :)